src/Microsoft.CodeAnalysis.BannedApiAnalyzers/UnitTests/RestrictedInternalsVisibleToAnalyzerTests.cs

```diff
@@ -0,0 +1,1931 @@
+        Public Function GetC1(Of U)() As C1(Of Object)
+            Return Nothing
+        End Function
+    End Class
+
+    Friend Class C3
+    End Class
+End Namespace";
+
+            var apiConsumerSource = @"
+Imports N1
+
+Class C2
+    Inherits C1(Of C3)
+
+    Private Sub M(ByVal c1 As C1(Of C3), ByVal c2 As C1(Of Object))
+        Dim unused1 = c2.GetC1(Of C3)()
+        Dim unused2 = c2.GetC1(Of C1(Of C3))()
+    End Sub
+End Class";
+
+            await VerifyBasicAsync(apiProviderSource, apiConsumerSource);
+        }
+
+        [Fact]
+        public async Task CSharp_IVT_RestrictedIVT_UsageInTypeArgument_Diagnostic()
+        {
+            var apiProviderSource = @"
+[assembly: System.Runtime.CompilerServices.InternalsVisibleTo(""ApiConsumerProjectName"")]
+[assembly: System.Runtime.CompilerServices.RestrictedInternalsVisibleTo(""ApiConsumerProjectName"", ""N2"")]
+
+namespace N1
+{
+    internal class C3 { }
+    internal class C4 { }
+    internal class C5 { }
+    internal class C6 { }
+
+}
+
+namespace N2
+{
+    internal class C1<T>
+    {
+        public C1<object> GetC1<U>() => null;
+    }
+}
+";
+
+            var apiConsumerSource = @"
+using N1;
+using N2;
+
+class C2 : C1<C3>
+{
+    void M(C1<C4> c1, C1<object> c2)
+    {
+        _ = c2.GetC1<C5>();
+        _ = c2.GetC1<C1<C6>>();
+    }
+}";
+
+            await VerifyCSharpAsync(apiProviderSource, apiConsumerSource,
+                GetCSharpResultAt(5, 15, "N1.C3", "N2"),
+                GetCSharpResultAt(7, 15, "N1.C4", "N2"),
+                GetCSharpResultAt(9, 22, "N1.C5", "N2"),
+                GetCSharpResultAt(10, 25, "N1.C6", "N2"));
+        }
+
+        [Fact]
+        public async Task Basic_IVT_RestrictedIVT_UsageInTypeArgument_Diagnostic()
+        {
+            var apiProviderSource = @"
+<Assembly: System.Runtime.CompilerServices.InternalsVisibleTo(""ApiConsumerProjectName"")>
+<Assembly: System.Runtime.CompilerServices.RestrictedInternalsVisibleTo(""ApiConsumerProjectName"", ""N2"")>
+
+Namespace N1
+    Friend Class C3
+    End Class
+    Friend Class C4
+    End Class
+    Friend Class C5
+    End Class
+    Friend Class C6
+    End Class
+End Namespace
+
+Namespace N2
+    Friend Class C1(Of T)
+        Public Function GetC1(Of U)() As C1(Of Object)
+            Return Nothing
+        End Function
+    End Class
+End Namespace";
+
+            var apiConsumerSource = @"
+Imports N1
+Imports N2
+
+Class C2
+    Inherits C1(Of C3)
+
+    Private Sub M(ByVal c1 As C1(Of C4), ByVal c2 As C1(Of Object))
+        Dim unused1 = c2.GetC1(Of C5)()
+        Dim unused2 = c2.GetC1(Of C1(Of C6))()
+    End Sub
+End Class";
+
+            await VerifyBasicAsync(apiProviderSource, apiConsumerSource,
+                GetCSharpResultAt(6, 20, "N1.C3", "N2"),
+                GetCSharpResultAt(8, 37, "N1.C4", "N2"),
+                GetCSharpResultAt(9, 35, "N1.C5", "N2"),
+                GetCSharpResultAt(10, 41, "N1.C6", "N2"));
+        }
```

Comment:
Consider adding a couple tests to cover extension method and using alias

Reply:
Done